soc/hps_soc.py

```diff
@@ -54,6 +55,72 @@
 SOC_DIR = os.path.dirname(os.path.realpath(__file__))
 
 
+class CfuCpuClockCtrl(Module):
+    """
+    A module that controls clocks between CFU and CPU so that power usage is
+    optimized.
+    """
+
+    def __init__(self, cfu_bus, cfu_cen, cpu_cen):
+        """Constructor
+
+        Args:
+            cfu_bus: bus between CFU and CPU
+            cfu_cen: clock enable signal for CFU
+            cpu_cen: clock enable signal for CPU
+        """
+        self.cfu_bus = cfu_bus
+        self.cfu_cen = cfu_cen
+        self.cpu_cen = cpu_cen
+
+        delay = Signal(max=10)
+
+        self.submodules.fsm = fsm = FSM()
+        # It's just an initial state, once left it's not entered again
+        fsm.act("RESET",
+            # Enable both CFU and CPU
+            self.cfu_cen.eq(1),
+            self.cpu_cen.eq(1),
+
+            NextValue(delay, delay + 1),
+
+            # After 10 cycles disable CFU and go to next state
```

Comment:
Where does the 10 cycles come from? I guess this is because the CFU logic does a little bit of work after reset, to get itself ready?

Reply:
Actually I've never checked that precisely until now. I just assumed that it requires some time to set up. After running some tests it looks like it doesn't need any setup time for most functions. In fact the only thing that fails without these additional cycles is a CFU ping test :) These results look like that because I did test reset executed by software which takes too many cycles from reset to next instruction to precisely check if CFU requires any setup time.

However I noticed that to make setup time relevant for CFU it would require to make `cmd.ready` synchronous which is not the case right now for some reason. Even if CFU is in reset state, CPU won't see it as inavailable if last CFU's state was `WAIT_CMD` since `cmd.ready` will be asserted whole time. Before that, I think there is no point in setting any setup time for CFU. 

I don't think it's something that would bother us right now. Issues don't come up for any tests in current state of CFU Playground and I had to force CFU reset right before calling `cfu_set()` to make ping test fail. If you have any case in hps-firmware that can be affected in some specific case by this issue, then it should be fixed I guess. Otherwise it might not be worth to do it.